src/main/java/net/sf/rails/util/RunGame.java

```diff
@@ -1,38 +1,54 @@
 package net.sf.rails.util;
 
+import java.awt.*;
+import java.awt.desktop.OpenFilesEvent;
+import java.awt.desktop.OpenFilesHandler;
 import java.io.File;
 
+import javax.swing.*;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
 import net.sf.rails.common.ConfigManager;
+import net.sf.rails.ui.swing.AutoSaveLoadDialog;
 import net.sf.rails.ui.swing.GameSetupController;
 
 public class RunGame {
 
+    private static final Logger log = LoggerFactory.getLogger(RunGame.class);
+
     public static void main(String[] args) {
+        final String[] fileName = { null };
 
-        // Initialize configuration
-        ConfigManager.initConfiguration(false);
-        
-        int nargs = 0;
         if (args != null && args.length > 0) {
-            nargs = args.length;
-            System.out.println("Number of args: "+nargs);
             for (String arg : args) {
                 System.out.println ("Arg: "+arg);
             }
+            fileName[0] = args[0];
+        }
+
+        if ( Desktop.isDesktopSupported() && Desktop.getDesktop().isSupported(Desktop.Action.APP_OPEN_FILE)) {
+            Desktop.getDesktop().setOpenFileHandler(e -> {
+                for ( File file : e.getFiles() ) {
+                    // open the file
+                    fileName[0] = file.toString();
+                    // break out as we only handle one game at a time...
+                    break;
+                }
+            });
```

Comment:
Is this executed in a blocking manner? 
If not this will not work, because you can't be sure that this is executed before the parts after the `if` condition

Reply:
On the Mac this executes immediately after the handler is installed, hence `fileName[0]` is set before the `if` is hit. I don't know about other platforms. And yes, this does work on the Mac; I've been using it for a couple of weeks now although it works much better now with your jpackage changes as the previous `dmg` creation required me to update the `Info.plist` manually with the file associations after I created the `dmg`.. hence making my build and testing process quite tedious. Finally wrote a shell script stringing together the commands and `patch` and can not ditch that

Reply:
I think that the `OpenFileHandler` is executed in the Swing thread and not in your normal application thread. Therefore it should be random that you got your expected result.
In addition an introduction in this topic from [Oracle](https://docs.oracle.com/javase/tutorial/deployment/selfContainedApps/fileassociation.html) suggests that this approach only works for Mac. For Linux and Windows you get the opened file as an input argument.

As a possible alternative it may be worth it to take a look at how this can be achieved with JavaFX, especially if we want to migrate more of our UI to JavaFX.

Reply:
Not sure how JavaFX would change our handling of receiving open events from the OS? I don't know much about JavaFX (in fact my only exposure is what we have in Rails) so any pointers would be appreciated for reading.

Either way, until we determine whether we would pursue anything JavaFX as an alternative, this approach does work for the bundled Mac app :)

Reply:
> Not sure how JavaFX would change our handling of receiving open events from the OS? I don't know much about JavaFX (in fact my only exposure is what we have in Rails) so any pointers would be appreciated for reading.

I also have no experience with JavaFX + file associations. A quick search led me to [this](https://stackoverflow.com/a/52606228) StackOverflow question. I think it may be a good starting point. 

> Either way, until we determine whether we would pursue anything JavaFX as an alternative

I agree. The part about a JavaFX solution was meant as a long-term goal.

> ..., this approach does work for the bundled Mac app

I still think that the `OpenFileHandler` is called from a separate thread, which makes this solution very unpredictable and unstable.

You could check this by printing the thread inside the `OpenFileHandler` and outside it by accessing `Thread.currentThread()`.

Reply:
Read the Stackoverflow link-- that may work for Linux and Windows based on the comments regarding how they are passed the file but will not work on the Mac as opening files are never passed as an argument.

Over the last several weeks of using this I've seen no random behavior-- it has *always* opened the file double clicked on.

Reply:
> Read the Stackoverflow link-- that may work for Linux and Windows based on the comments regarding how they are passed the file but will not work on the Mac as opening files are never passed as an argument.

I don't know if this will work on Mac. It is just meant as a starting point. The long-term problem with the `Desktop` solution is that it is `awt` so if we want to move from `swing`/`awt` to JavaFX long-term we will either need to replace this with something that purely JavaFX or we accept a mix of JavaFX and `swing`/`awt`.

But this is not a question we must face now, it is simply meant as a topic which we can think about in the future.

> Over the last several weeks of using this I've seen no random behavior-- it has always opened the file double clicked on.

I can understand that this has always worked for you. But that an issue didn't have a visible impact does not mean that it is not there right? To exclude the possibility that the `OpenFileHandler` is executed from a separate thread I would really appreciate it if you could test whether the thread that calls the `OpenFileHandler` is equal to the one that executes:

https://github.com/Rails-18xx/Rails/blob/27cc1391e455f1e10bdbf3293e78de4e4a1b4614/src/main/java/net/sf/rails/util/RunGame.java#L42-L52

If this is the case then everything as it should be, otherwise we know that we can face a race condition.

Reply:
As I believe I mentioned previously-- this code has already been changed to add another feature so am unable to spend time to backtrack and try to test for an edge case that you may believe is there that will be a non-issue once this PR has been merged and I can submit the following PR

Reply:
But DO welcome your review on the following PR for similar....

Reply:
Ok